### PR TITLE
EventStore repository to allow connection factory to be passed

### PR DESCRIPTION
**Pull request by ninjaboy**
This change introduces connection factory method as a dependency to the EventStore repository instead of connection object to allow more flexibility with regards to how connection object is resolved. (e.g. to allow connection polling, re-connection and monitoring)

Work done
---

  1. Major changes

Changed 	`EventStoreRepository` constructor from
        `public EventStoreRepository(IHoldAllConfiguration configs, IEventStoreConnection connection)`
to	        
        `public EventStoreRepository(IHoldAllConfiguration configs, Func<IEventStoreConnection> connectionFactory)`

  2. Important things to review

`EventStoreRepository` as the method to create new session was changed

Integrity
---

- [x] I have had a look on the PR preview for obvious whitespace\formatting issues before actually openned this PR
- [x] I have run unit tests
- [x] I have run EventStore integration tests

### CODE REVIEW

File: src/BullOak.Repositories.EventStore/EventstoreRepository.cs [hunk 1]
```diff
@@ -11,26 +11,43 @@ public class EventStoreRepository<TId, TState> : IStartSessions<TId, TState>
...
+                connection = connectionFactory();
+            }
+            catch (Exception ex)
+            {
+                throw new RepositoryUnavailableException("Couldn't connect to the EvenStore repository. See InnerException for details", ex);
+            }
+
+            if (connection == null)
+            {
+                throw new RepositoryUnavailableException("Couldn't connect to the EvenStore repository. See InnerException for details", new ArgumentNullException(nameof(connection)));
```
[the remark this answers is not in the document]
Comment (ninjaboy) [reply src/BullOak.Repositories.EventStore/EventstoreRepository.cs]: Valid point

[the remark this answers is not in the document]
Comment (ninjaboy) [reply src/BullOak.Repositories.EventStore/EventstoreRepository.cs]: Removed InnerException at all

**Comment by ninjaboy** [comment]
@iblazhko , thanks for the spot Ivan, I have

File: src/BullOak.Repositories.EventStore/EventstoreRepository.cs [hunk 2]
```diff
@@ -9,28 +9,74 @@
...
+            {
+                CleanupConnection(connection);
+                throw;
+            }
 
             return session;
         }
 
-        public async Task<bool> Contains(TId selector)
+        private void CleanupConnection(IEventStoreConnection connection)
```
Comment (iblazhko) [inline src/BullOak.Repositories.EventStore/EventstoreRepository.cs]: This cleanup function could potentially be shared between EventStoreRepository and EventStoreSession

Comment (ninjaboy) [reply src/BullOak.Repositories.EventStore/EventstoreRepository.cs]: Done

**Comment by ninjaboy** [comment]
@skleanthous , Shall explore the following approach then:
```
EventStoreRepository repository; //injected via ctor
void Operation(string id){
   var connection = GetConnection(); //client decides whether it's from pool or a new one or whatever
   using(repository.BeginSessionForId(connection, id)){
       //do stuff...
   }
   //now I decide what to do with the connection
   connection.Close();
}
```

**Comment by skleanthous** [comment]
@ninjaboy Something like `connectionFactory.GetSingleInstance()` vs `connectionFactory.GetTransient()`? If so then yeah, that's perfect.

**Comment by ninjaboy** [comment]
@skleanthous , `GetConnection()` -> This is up for a client to decide. They may choose whatever strategy to manage the connection (e.g. reuse between sessions, or open close or whatever, the session only requires the connection to be open.

However, looking at the interface that repository currently implements I won't be able to pass arbitrary parameters to the session. Because I have to stick to:

```
    public interface IStartSessions<in TEntitySelector, TState>
    {
        Task<IManageSessionOf<TState>> BeginSessionFor(TEntitySelector selector, bool throwIfNotExists);
        Task Delete(TEntitySelector selector);
        Task<bool> Contains(TEntitySelector selector);
    }
```

**Comment by ninjaboy** [comment]
I must say that I am really confused so far with how the issue needs to be resolved.
Here are some links that I am trying to follow to figure out the possible solution:
https://github.com/EventStore/EventStore/issues/1584
https://github.com/EventStore/EventStore/issues/1732
https://groups.google.com/forum/#!topic/event-store/nkgzXpy56NQ
https://groups.google.com/forum/#!topic/event-store/xu1GQzZtCiM
https://groups.google.com/forum/#!topic/event-store/z0RWBnBesoU

Any suggestion on how we shall handle the problem with connection object going into the Disposed state are welcome.

**Comment by ninjaboy** [comment]
https://groups.google.com/forum/#!topic/event-store/nkgzXpy56NQ
This post is particularly helpful and in the end it mentions the change introduced in v4.1.1 that introduces a function `SetMaxQueueTimeout` that looks ismilar to what we need. However .net core version of event store client exists only for version 4.1.0.23 and the new call is introduced in v 4.1.1.

**Comment by skleanthous** [comment]
Will be done in a future release

**Pull request closed without merging**